### PR TITLE
implement `make_quadratic_cqm`

### PR DESCRIPTION
**Pull request by mhlr**
Build on #1033

### CODE REVIEW

**Comment by arcondello** [comment]
Can you rebase this once you get the chance?

**Comment by arcondello** [comment]
I just noticed that CI is not running. Not sure what's happening there...

**Comment by arcondello** [comment]
Closing and opening seems to have fixed the CI. Looks like #1033 introduced some doctest failures. We can maybe resolve them as part of this PR?

**Comment by mhlr** [comment]
> 
> 
> Closing and opening seems to have fixed the CI. Looks like #1033 introduced some doctest failures. We can maybe resolve them as part of this PR?

I could not figiure why it was runni9ng on this or #1033

**Comment by mhlr** [comment]
> 
> 
> I just noticed that CI is not running. Not sure what's happening there...

It has not been running on this or #1033

**Comment by mhlr** [comment]
> 
> 
> Closing and opening seems to have fixed the CI. Looks like #1033 introduced some doctest failures. We can maybe resolve them as part of this PR?

Done

File: dimod/higherorder/utils.py [hunk 1]
```diff
@@ -174,6 +175,94 @@ def reduce_binary_polynomial(poly: BinaryPolynomial) -> Tuple[
...
+    vartype = vartype or poly.vartype
+    poly = _init_binary_polynomial(poly, vartype)
+    variables = set().union(*poly)
+    reduced_terms, constraints = reduce_binary_polynomial(poly)
+
+    def var(x):
+        return BinaryQuadraticModel({x: 1.0}, {}, 0.0, vartype)
+
+    for (u, v), p in constraints:
+        cqm.add_constraint( var(u)*var(v) - var(p)  == 0, label = f'var__{u}*var_{v} == var_{p}')
```
Comment (arcondello) [inline dimod/higherorder/utils.py:258]: why not just `label=f'{u!r}*{v!r}=={p!r}`? might be confusing to later look for `var__{u}` in the `.variables`.

Comment (mhlr) [reply dimod/higherorder/utils.py:258]: > 
> 
> why not just `label=f'{u!r}*{v!r}=={p!r}`? might be confusing to later look for `var__{u}` in the `.variables`.

I thought about this. Main reaso for the `var_` is that the equation will look weird if the variables are numbers. Maybe that is ok. or maybe use something like `v{u}` or `x{u}`.. Not sure. I do not have strong feelings about this.

Comment (arcondello) [reply dimod/higherorder/utils.py:258]: Ok, in that case I think I am comfortable with the weird integer case. It is just a label after all and should be pretty clear from context.

Comment (mhlr) [reply dimod/higherorder/utils.py:257]: maybe wrap quotes?

**Comment by mhlr** [comment]
> 
> 
> > I just noticed that CI is not running. Not sure what's happening there...
> 
> It has not been running on this or #1033

Looks like CircleCI is off again on this PR

**Comment by arcondello** [comment]
> Looks like CircleCI is off again on this PR

yes - whenever you make new commits it turns back off until I close and re-open. You have unfollowed the project on your personal circleci?

**Comment by mhlr** [comment]
> 
> 
> > Looks like CircleCI is off again on this PR
> 
> yes - whenever you make new commits it turns back off until I close and re-open. You have unfollowed the project on your personal circleci?

image

**Comment by arcondello** [comment]
Had you made this PR before unfollowing? If yes, maybe pushing to a new branch and making a new PR will fix it. :shrug:

**Comment by mhlr** [comment]
> 
> 
> Had you made this PR before unfollowing? If yes, maybe pushing to a new branch and making a new PR will fix it. 🤷

Ok. I do not think I ever follwed in the first place though. Not sure I ever logged into my CircleCI acct. Did not realize I had one.

**Comment by arcondello** [comment]
> Ok. I do not think I ever follwed in the first place though. Not sure I ever logged into my CircleCI acct. Did not realize I had one.

Eh, no idea then. Whatever, I can just close/open the PR whenever it's needed.

**Comment by mhlr** [comment]
> 
> 
> > Closing and opening seems to have fixed the CI. Looks like #1033 introduced some doctest failures. We can maybe resolve them as part of this PR?
> 
> Done

I misunderstood you initial comment here. I do not understand these doctest failures. What is  it that is actually failing. Why am I not seeing unint test failures locally?

**Comment by arcondello** [comment]
> I misunderstood you initial comment here. I do not understand these doctest failures. What is it that is actually failing. Why am I not seeing unint test failures locally?

You're running
```bash
make -C docs/ build doctest
```
or the equivalent?

**Comment by mhlr** [comment]
> 
> 
> > I misunderstood you initial comment here. I do not understand these doctest failures. What is it that is actually failing. Why am I not seeing unint test failures locally?
> 
> You're running
> 
> ```shell
> make -C docs/ build doctest
> ```
> 
> or the equivalent?

` python -m unittest`

**Comment by arcondello** [comment]
> ` python -m unittest`

that runs the unittests, not the doctests. Could look into adding doctests to unittests but right now they are not the same thing.

**Comment by mhlr** [comment]
> 
> 
> > ` python -m unittest`
> 
> that runs the unittests, not the doctests. Could look into adding doctests to unittests but right now they are not the same thing.

Guess I never used doctests. Gotta read up on them. ATM I get

```
(base) daniel@nb-000-2831:~/dimod$ make -C docs/ build doctest
make: Entering directory '/home/daniel/dimod/docs'

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/home/daniel/anaconda3/lib/python3.8/site-packages/sphinx/config.py", line 328, in eval_config_file
    exec(code, namespace)
  File "/home/daniel/dimod/docs/conf.py", line 126, in <module>
    import sphinx_rtd_theme
ModuleNotFoundError: No module named 'sphinx_rtd_theme'

make: *** [Makefile:25: build] Error 2
make: Leaving directory '/home/daniel/dimod/docs'
(base) daniel@nb-000-2831:~/dimod$
```                                        `

**Comment by arcondello** [comment]
```bash
pip install -r docs/requirements.txt
make -C docs/ build doctest
```

File: releasenotes/notes/make_quadratic_cqm-bbcc2957eebbfad9.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    added make_quadratic_cqm - like make quadratic, but returns a CQM treating  auxillary variable constraints as proper constraints instead of adding them to the objective
```
Comment (arcondello) [inline releasenotes/notes/make_quadratic_cqm-bbcc2957eebbfad9.yaml:4]: ``` ``make_quadratic_cqm() `` ```. Also capitalization, some extra spaces. Also not sure what the word "proper" is adding.

File: dimod/higherorder/utils.py [hunk 3]
```diff
@@ -110,8 +112,8 @@ def reduce_binary_polynomial(poly: BinaryPolynomial) -> Tuple[
         ([(term, bias)*], [((orig_var1, orig_var2), aux_var)*])
 
     Example:
-        >>> poly = BinaryPolynomial({(0,): -1, (1,): 1, (2,): 1.5, (0, 1): -1, (0, 1, 2): -2}, dimod.BINARY)
-        >>> reduce_binary_polynomial(poly)
+        >>> poly = BinaryPolynomial({(0,): -1, (1,): 1, (2,): 1.5, (0, 1): -1, (0, 1, 2): -2}, dimod.BINARY) # doctest: +SKIP
```
Comment (arcondello) [inline dimod/higherorder/utils.py:115]: I would prefer to skip as little as possible. I think just `poly = dimod.BinaryPolynomail(...` would also avoid failures.

Comment (mhlr) [reply dimod/higherorder/utils.py:115]: Will trywith qualification. BinaryPolynomial is imported explicitly though.

Comment (mhlr) [reply dimod/higherorder/utils.py:115]: So the qualification does fix the first line of the doctest, but running
```
(cd docs; sphinx-build -b doctest . build/doctest)
```
locally gives me:
```
Document: reference/generated/dimod.higherorder.utils.reduce_binary_polynomial
------------------------------------------------------------------------------
**********************************************************************
File "../dimod/higherorder/utils.py", line ?, in default
Failed example:
    reduce_binary_polynomial(poly)
Exception raised:
    Traceback (most recent call last):
      File "/home/daniel/anaconda3/lib/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest default[1]>", line 1, in <module>
        reduce_binary_polynomial(poly)
    NameError: name 'reduce_binary_polynomial' is not defined
**********************************************************************
1 items had failures:
   1 of   2 in default
2 tests in 1 items.
1 passed and 1 failed.
***Test Failed*** 1 failures.
```

Comment (arcondello) [reply dimod/higherorder/utils.py:115]: Ok, I was following up with @JoelPasvolsky , let's just leave it as-is for now. Can circle back around later.

Comment (mhlr) [reply dimod/higherorder/utils.py:115]: Ok I will only skip the second line then

**Review by arcondello — APPROVED** [review]

**Comment by arcondello** [comment]
@mhlr , ready to merge?

**Comment by mhlr** [comment]
> 
> 
> @mhlr , ready to merge?

unit tests and doctests pass. I fixed the labels and the release note text. I think that is all there was

**Pull request merged**